sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala

```diff
@@ -107,46 +109,65 @@ abstract class StringRegexExpression extends BinaryExpression
       true
       > SELECT '%SystemDrive%/Users/John' _FUNC_ '/%SystemDrive/%//Users%' ESCAPE '/';
       true
+      > SELECT _FUNC_('_Apache Spark_', '__%Spark__', '_');
+      true
   """,
   note = """
     Use RLIKE to match with standard regular expressions.
   """,
   since = "1.0.0")
 // scalastyle:on line.contains.tab
-case class Like(left: Expression, right: Expression, escapeChar: Char)
-  extends StringRegexExpression {
+case class Like(str: Expression, pattern: Expression, escape: Expression)
+  extends TernaryExpression with StringRegexExpression {
 
-  def this(left: Expression, right: Expression) = this(left, right, '\\')
+  def this(str: Expression, pattern: Expression) = this(str, pattern, Literal("\\"))
+
+  override def inputTypes: Seq[DataType] = Seq(StringType, StringType, StringType)
+  override def children: Seq[Expression] = Seq(str, pattern, escape)
+
+  private lazy val escapeChar: Char = if (escape.foldable) {
```

Comment:
We need `lazy` here? I personally think its better to check error conditions as soon as possible.

Reply:
When I remove `lazy`, I get the exception:
```
makeCopy, tree: str#13580 LIKE pattern#13581 ESCAPE '@'
org.apache.spark.sql.catalyst.errors.package$TreeNodeException: makeCopy, tree: str#13580 LIKE pattern#13581 ESCAPE '@'
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.catalyst.trees.TreeNode.makeCopy(TreeNode.scala:435)
	at org.apache.spark.sql.catalyst.trees.TreeNode.mapChildren(TreeNode.scala:408)
...
Caused by: org.apache.spark.sql.AnalysisException: The 'escape' parameter must be a string literal.;
	at org.apache.spark.sql.catalyst.expressions.Like.<init>(regexpExpressions.scala:135)
```
See `escape` is `PrettyAttribute`:
<img width="721" alt="Screen Shot 2020-01-26 at 19 51 34" src="https://user-images.githubusercontent.com/1580697/73138597-8716ef80-4075-11ea-9cb5-b92baef6d4c5.png">

Reply:
I got the error when I run new test in `DataFrameFunctionsSuite`. PrettyAttribute is not foldable.

Reply:
Ur, ok. Thanks for the check. The current one looks ok to me.